File: common/src/main/java/dev/obscuria/tooltips/client/filter/ItemFilter.java
package dev.obscuria.tooltips.client.filter;

import com.mojang.serialization.Codec;
import dev.obscuria.tooltips.registry.TooltipsRegistries;
import net.minecraft.world.item.ItemStack;

import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;

public interface ItemFilter {

    Codec<ItemFilter> CODEC = TooltipsRegistries.ITEM_FILTER_TYPE.byNameCodec().dispatch(ItemFilter::codec, Function.identity());

    Codec<? extends ItemFilter> codec();

    boolean test(ItemStack stack);

    static void bootstrap(BiConsumer<String, Supplier<Codec<? extends ItemFilter>>> registrar) {

        registrar.accept("always", () -> AlwaysFilter.CODEC);
        registrar.accept("never", () -> NeverFilter.CODEC);
        registrar.accept("all_of", () -> AllOfFilter.CODEC);
        registrar.accept("any_of", () -> AnyOfFilter.CODEC);
        registrar.accept("none_of", () -> NoneOfFilter.CODEC);
        registrar.accept("item", () -> ItemOrTagFilter.CODEC);
        registrar.accept("mod", () -> ModFilter.CODEC);
        registrar.accept("enchantment", () -> EnchantmentFilter.CODEC);
        registrar.accept("rarity", () -> RarityFilter.CODEC);
        registrar.accept("nbt", () -> NbtFilter.CODEC);
        registrar.accept("property", () -> PropertyFilter.CODEC);
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/ItemFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.tags.TagKey
import net.minecraft.util.ExtraCodecs
import net.minecraft.util.ExtraCodecs.TagOrElementLocation
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack

class ItemFilter(
    val tags: List<TagKey<Item>>,
    val items: List<Item>
) : IItemFilter {

    override fun codec(): Codec<ItemFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return items.contains(stack.item) || tags.any { stack.`is`(it) }
    }

    fun pack(): List<TagOrElementLocation> {
        val packedTags = tags.map { TagOrElementLocation(it.location, true) }
        val packedItems = items.map { TagOrElementLocation(it.builtInRegistryHolder().key().location(), false) }
        return packedTags + packedItems
    }

    companion object {

        val CODEC: Codec<ItemFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                ExtraCodecs.TAG_OR_ELEMENT_ID.listOf()
                    .fieldOf("items")
                    .forGetter(ItemFilter::pack)
            ).apply(codec, ::unpack)
        }

        fun unpack(packed: List<TagOrElementLocation>): ItemFilter {
            val tags = mutableListOf<TagKey<Item>>()
            val items = mutableListOf<Item>()
            packed.forEach {
                if (it.tag) tags += TagKey.create(Registries.ITEM, it.id)
                else items += BuiltInRegistries.ITEM.get(it.id)
            }
            return ItemFilter(tags, items)
        }
    }
}